plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -781,28 +720,8 @@ void producer_plugin::plugin_initialize(const boost::program_options::variables_
       const std::vector<std::string> key_spec_pairs = options["signature-provider"].as<std::vector<std::string>>();
       for (const auto& key_spec_pair : key_spec_pairs) {
          try {
-            auto delim = key_spec_pair.find("=");
-            EOS_ASSERT(delim != std::string::npos, plugin_config_exception, "Missing \"=\" in the key spec pair");
-            auto pub_key_str = key_spec_pair.substr(0, delim);
-            auto spec_str = key_spec_pair.substr(delim + 1);
-
-            auto spec_delim = spec_str.find(":");
-            EOS_ASSERT(spec_delim != std::string::npos, plugin_config_exception, "Missing \":\" in the key spec pair");
-            auto spec_type_str = spec_str.substr(0, spec_delim);
-            auto spec_data = spec_str.substr(spec_delim + 1);
-
-            auto pubkey = public_key_type(pub_key_str);
-
-            if (spec_type_str == "KEY") {
-               my->_signature_providers[pubkey] = make_key_signature_provider(private_key_type(spec_data));
-            } else if (spec_type_str == "KEOSD") {
-               my->_signature_providers[pubkey] = make_keosd_signature_provider(my, spec_data, pubkey);
-            }
-#ifdef __APPLE__
-            else if (spec_type_str == "SE") {
-               my->_signature_providers[pubkey] = make_se_signature_provider(pubkey);
-            }
-#endif
+            auto [pubkey, provider] = app().get_plugin<signature_provider_plugin>().signature_provider_for_specification(key_spec_pair);
```

Comment:
I would make this ```const auto&```.  The compiler should optimize what you have, but if not then it is two non-trivial copies, and one more copy for the provider.